include/xlnt/utils/date.hpp

```diff
@@ -49,21 +50,54 @@ struct XLNT_API date
     static date from_number(int days_since_base_year, calendar base_date);
 
     /// <summary>
-    /// Constructs a data from a given year, month, and day.
+    /// Constructs a date from a given year, month, and day.
     /// </summary>
     date(int year_, int month_, int day_);
 
     /// <summary>
-    /// Return the number of days between this date and base_date.
+    /// Constructs an empty date (a call to is_null() will return true).
+    /// </summary>
+    date() = default;
+
+    /// <summary>
+    /// Return the number of days between this date and base_date. The date could be in an empty/invalid state, so you may want to call is_null() first!
+    /// If this function is called when having an empty/invalid state, an exception will be thrown.
     /// </summary>
     int to_number(calendar base_date) const;
 
     /// <summary>
     /// Calculates and returns the day of the week that this date represents in the range
     /// 0 to 6 where 0 represents Sunday.
+    /// Returns -1 if the weekday could not be determined.
```

Comment:
If you would like to avoid the magic number, we could also throw an exception instead, e.g. `xlnt::unsupported` . However, I'm not sure which option I prefer, so you can choose freely.

Reply:
Personally, for my own C++ projects, I prefer to avoid exceptions whenever possible, because there's almost always a better solution (except if you work with some library that requires exception handling). Also, in performance-critical code, catching exceptions frequently reduces performance considerably (probably not important here, but it's worth to mention). An alternative to exception handling could either be a `bool` return value specifying whether the value is valid, or even better, `std::optional` if C++17 is available (or, in our case, `xlnt::optional`). I prefer instead to do the following:

- Use exceptions, if possible, only when having absolutely unrecoverable errors (together with an unhandled exception handler).
- Otherwise, use exceptions only if required by some library and catch the exception as soon as possible, WITHOUT propagating it.
- Otherwise use `bool`s or `optional`s for signaling error cases. For more complex errors, I prefer `std::variant` (to have either a result or an error) if C++17 support is available, or even better, `std::expected` if C++23 support is available.

Also, if stated very clearly, I also have nothing against using exceptions if the programmer messed up badly, e.g. accessed the value of an `std::optional` or `xlnt::optional`, or accessed invalid indices when using `std::vector::at()`. This effectively means that it is NOT necessary to write `try` / `catch` all over the code in order to write correct code. I think this is important because C++ does not enforce handling all exceptions when calling a function that could throw, so it requires everyone to read the documentation (if it exists at all; and even then, many people unfortunately don't read the documentation).

However, I do agree that in this specific case, having a behavior similar to `std::optional` or `xlnt::optional` in the sense that access to invalid values is prevented using exceptions is definitely much better than allowing unrestricted access to invalid values. However, what I personally want to avoid is having to write `try` / `catch` all over the code due to things the programmer cannot control, e.g. a failed internal call to `detail::localtime_safe` or `std::mktime`.

In conclusion, while I really dislike using a magic number, using an exception in this case provides no advantages but a huge disadvantage, since it requires exception handling everywhere, making the situation even worse.

The only proper solution, in my opinion, is to return an `xlnt::optional` - this way, the programmer can see right away that the call can fail, even without reading the documentation. This also applies to `date::today()`, `datetime::today()` and `datetime::now()` (although, one could argue that it's unlikely to not be able to get today's date and/or time, so maybe using an `optional` here would be exaggerated). And now I even got to a point where I can argue that, when returning an `optional` in this case(s), our internal variable `_is_null` is not even necessary anymore, as the structs will always only contain a valid state (at least in the sense that we're not checking any ranges for the year, month, day etc.).

Of course, all of this is not possible without changing a public interface, so not doable if we want to release a minor version.

Either way, I still think that getting rid of the public members in favor of the getters is a very good idea (e.g. so that the struct can only be changed during creation). I only explained the whole thing above to prove why I think that a slightly different solution would be better, but our current solution is still good enough for now, I think.

Reply:
Ok. This rule of thumb is fine for me.

Removing `_is_null` in favor of return an `xlnt::optional` may be indeed an option for a major upgrade. In that case, it may be useful to make the constructor private too and use a factory method so we can signal a failed construction too (if desired).

Reply:
Maybe we should already create a v2.0.0 milestone so we can keep track of the changes we want to implement once we release a new major version.

Reply:
Agreed, I think this would be a great idea :+1: